libs/template/config.go

```diff
@@ -92,63 +92,102 @@ func (c *config) assignDefaultValues(r *renderer) error {
+	name := property.Name
+	schema := property.Schema
 
-		// Config already has a value assigned
-		if _, ok := c.values[name]; ok {
-			continue
-		}
+	// Compute description
+	description, err := r.executeTemplate(schema.Description)
+	if err != nil {
+		return err
+	}
 
-		// Compute default value to display by converting it to a string
-		var defaultVal string
-		var err error
-		if property.Default != nil {
-			defaultValRaw, err := property.DefaultString()
-			if err != nil {
-				return err
-			}
-			defaultVal, err = r.executeTemplate(defaultValRaw)
-			if err != nil {
-				return err
-			}
-		}
+	// List of options to display to user
+	options, err := schema.EnumStringSlice()
+	if err != nil {
+		return err
+	}
+
+	// Get user input
+	userInput, err := cmdio.AskSelect(c.ctx, description, options)
+	if err != nil {
+		return err
+	}
+
+	// Convert user input string back to a value
+	c.values[name], err = schema.ParseString(userInput)
+	if err != nil {
+		return err
+	}
+
+	// Validate the partial config based on this update
+	return c.schema.ValidateInstance(c.values)
+}
+
+func (c *config) promptText(property jsonschema.Property, r *renderer, numRetries int) error {
+	name := property.Name
+	schema := property.Schema
+
+	// Compute description
+	description, err := r.executeTemplate(schema.Description)
+	if err != nil {
+		return err
+	}
+
+	// Compute default value
+	defaultRaw, err := schema.DefaultString()
+	if err != nil {
+		return err
+	}
+	defaultVal, err := r.executeTemplate(defaultRaw)
+	if err != nil {
+		return err
+	}
 
-		description, err := r.executeTemplate(property.Description)
+	for i := 0; i < numRetries; i++ {
+		// Get user input. Parse the string back to a value
+		userInput, err := cmdio.Ask(c.ctx, description, defaultVal)
 		if err != nil {
 			return err
 		}
 
-		// Get user input by running the prompt
-		var userInput string
-		if property.Enum != nil {
-			// convert list of enums to string slice
-			enums, err := property.EnumStringSlice()
-			if err != nil {
-				return err
-			}
-			userInput, err = cmdio.AskSelect(c.ctx, description, enums)
-			if err != nil {
-				return err
-			}
-		} else {
-			userInput, err = cmdio.Ask(c.ctx, description, defaultVal)
-			if err != nil {
-				return err
-			}
-		}
+		// Convert user input string back to a Go value
+		c.values[name], err = schema.ParseString(userInput)
 
-		// Convert user input string back to a value
-		c.values[name], err = property.ParseString(userInput)
+		// Error parsing user input. Retry if not last attempt
+		if err != nil && i != numRetries-1 {
+			cmdio.LogString(c.ctx, fmt.Sprintf("Validation failed: %s", err.Error()))
+			continue
+		}
 		if err != nil {
 			return err
 		}
 
-		// Validate the partial config based on this update
-		if err := c.schema.ValidateInstance(c.values); err != nil {
+		// Validate the partial config based on this new value by the user
+		err = c.schema.ValidateInstance(c.values)
```

Comment:
This is duplicated in the other prompt function.

Reply:
Not really duplicated right? For `promptText` this is wrapped in a retry loop.

Reply:
So for `promptSelect` we never retry if it fails?

Reply:
I believe there's no reason to retry `promptSelect` atm. The reason we would like to retry `promptText` is because it's likely the user entered value is of the wrong type or does not match the regex pattern required.

In case of `promptSelect`, An error likely implies something went wrong with the prompting itself, rather than having to do something with the value a user selected.

Reply:
Done. Moved retry loop to the `promptForValues` function. It's probably good enough to retry on all validation errors. WDYT?